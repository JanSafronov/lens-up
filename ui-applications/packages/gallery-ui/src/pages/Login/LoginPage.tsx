import { ChangeEvent, FormEvent, useState } from "react";
import LoginForm from "./components/LoginForm"
import { useLoginToGalleryMutation } from "../../services/GalleryApi";
import { useNavigate } from "react-router-dom";

const LoginPage = () => {
    const navigate = useNavigate();
    const [enterCode, setEnterCode] = useState<string>();

    const [loginToGallery] = useLoginToGalleryMutation();

    const handleFormSubmit = (event: FormEvent) => {
        event.preventDefault()
        // TODO: Add validation
        if (!enterCode) {
            return;
        }

        loginToGallery({ enterCode: enterCode })
            .unwrap()
            .then((payload) => {
                navigate(`/gallery/${payload.galleryId}`)
            })
    }

    const handleEnterCodeInputChange = (event: ChangeEvent<HTMLInputElement>) => {
        // TODO: Validation
        const enterCodeValue = event.target.value;
        setEnterCode(enterCodeValue);
    };

    return (
        <div className="min-h-screen bg-black">
            <div className="container mx-auto flex flex-wrap flex-col md:flex-row justify-center h-screen items-center">
                <LoginForm handleFormSubmit={handleFormSubmit} enterCode={enterCode} handleEnterCodeInputChange={handleEnterCodeInputChange} />
            </div>
        </div>
    )
}

export default LoginPage