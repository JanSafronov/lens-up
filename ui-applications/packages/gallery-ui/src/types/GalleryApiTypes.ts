export interface LoginToGalleryPayload {
    enterCode: string,
}

export interface LoginToGalleryResponse {
    enterCode: string,
    galleryId: string,
    qrCodeUrl: string
}

export interface PhotoItem {
    id: string
    url: string
    authorName: string
    wishesText: string
}