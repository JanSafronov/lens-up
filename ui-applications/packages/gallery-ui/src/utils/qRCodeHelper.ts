import { LocalStorageKeys } from "./constants"

export const saveQRCode = (qRCodeUrl: string) => {
    localStorage.setItem(LocalStorageKeys.QRCodeUrl, qRCodeUrl)
}

export const getQRCodeUrl = () => {
    const url = localStorage.getItem(LocalStorageKeys.QRCodeUrl)
    if (!url) {
        console.error("QR code url not found in local storage.")
        return
    }

    return url
}