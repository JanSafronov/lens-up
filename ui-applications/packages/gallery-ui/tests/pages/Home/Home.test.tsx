import { cleanup, fireEvent, getAllByAltText, getByAltText, getByText, render, screen, waitFor } from '@testing-library/react'
import { expect, test, describe, beforeAll, beforeEach ,afterEach, afterAll } from 'vitest';
import { setupServer } from 'msw/node';
import { loginSubmit } from '../Login/Login.tsx';
import { getQRCodeUrl } from '../../../src/utils/qRCodeHelper.ts';
import { HttpResponse, http } from 'msw';
import { LoginToGalleryResponse } from '../../../src/types/GalleryApiTypes.ts';

let handlers = [
  http.post(`${import.meta.env.VITE_GALLERY_SERVICE_URL}/login`, async () => {
      let response: LoginToGalleryResponse = { enterCode: '0', galleryId: '0', qrCodeUrl: 'QRCodeUrl' };
      let res = await HttpResponse.json(response);
      return res;
  })
]

let asFragment: () => DocumentFragment;
let baseElement: HTMLElement;

/**
* @vitest-environment jsdom
*/
describe("Home page with no photos", async () => {
  const server = setupServer(...handlers);

  beforeAll(async () => {
    server.listen();
    ({ asFragment, baseElement } = await loginSubmit());
  })
  beforeEach(async () => waitFor(() => {
    expect(global.window.location.pathname).contains("/gallery/0");
  }))
  afterEach(cleanup);  
  afterAll(() => {
      server.close()
  })

  test('No photo hub must show qr code card in home page', async () => {
    const firstRender = asFragment();

    expect(firstRender).toMatchSnapshot();
  });

  test('No photo hub must have qr-code url in image source', async () => {
    const firstRender = asFragment();

    expect(getByAltText(baseElement, "QR code").getAttribute("src")).toEqual(getQRCodeUrl());

    expect(firstRender).toMatchSnapshot();
  });
});

describe("Home page with photos", async () => {
  handlers.concat(http.post(`${import.meta.env.VITE_GALLERY_SERVICE_URL}/hubs/gallery`, async (galleryId) => {
    let response: LoginToGalleryResponse = { enterCode: '0', galleryId: '0', qrCodeUrl: 'QRCodeUrl' };
    let res = await HttpResponse.json(response);
    return res;
  }))

  const server = setupServer(...handlers);

  beforeAll(async () => {
    server.listen();
    ({ asFragment, baseElement } = await loginSubmit());
  })
  beforeEach(async () => waitFor(() => {
    expect(global.window.location.pathname).contains("/gallery/0");
  }))
  afterEach(cleanup);  
  afterAll(() => {
      server.close()
  })

  test('Photo gallery must be shown', () => {
    const firstRender = asFragment();

    expect(firstRender).toMatchSnapshot();
  });
})