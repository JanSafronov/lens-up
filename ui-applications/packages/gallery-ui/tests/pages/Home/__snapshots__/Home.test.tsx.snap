// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home page with photos > Photo gallery must be shown 1`] = `
<DocumentFragment>
  <div
    class="min-h-screen bg-black"
  >
    <div
      class="absolute top-1/2 left-1/2 transform -translate-x-1/2"
    >
      <figure>
        <img
          alt="QR code"
          class="rounded-lg w-64 h-64"
          src="QRCodeUrl"
        />
        <figcaption
          class="font-bold text-base mb-2 text-center text-white"
        >
          Scan this and join.
        </figcaption>
      </figure>
    </div>
  </div>
</DocumentFragment>
`;
